projects/convex-hull/index.md

```diff
@@ -42,8 +42,16 @@ problem and the different algorithms to solve it.
 
 ## Testing
 
-Verify that the actual output matches the expected output. See the
-[requirements][3] section for an example of the expected output.
+The following table contains various test cases that you can use to
+verify the correctness of your solution; please note that different algorithms
+could produce the output starting from a different point, and/or in the opposite direction:
```

Comment:
I just noticed you raised a great point here! Since the point of testing is that we have something we can replicate, is there any way to specify an order of the output (e.g. smallest x first)? Or perhaps when we test, we just have to make sure all the points exist and order doesn't matter. @auroq, thoughts?

Reply:
That's a bit difficult to answer since, as I said, it depends on the algorithm used. Jarvis' algorithm (the one I used in my code) starts from the leftmost point (min x) while Graham's, for example, starts from the bottom one (min y - https://en.wikipedia.org/wiki/Graham_scan).

One naive way of testing would be to verify if all output points appear in the expected result ignoring the actual position, and just trusting that the algorithm was implemented correctly. Otherwise, I'd do something like this:

- take the expected result as an array of points and add a duplicate to it (A)
_[(100, 220)(240, 20)(180, 120)] => [(100, 220)(240, 20)(180, 120)(100, 220)(240, 20)(180, 120)]_
- reverse the initial array, and duplicate it as well (B)
_[(100, 220)(240, 20)(180, 120)] => [(180, 120)(240, 20)(100, 220)(180, 120)(240, 20)(100, 220)]_
- take the algorithm output as an array of points and, if either A or B contains the full sequence of points as the output, the test is validated; the idea is that by duplicating the expected points' sequence you are able to test the output around the end of the convex hull polygon without any weird index logic, and allowing for the output to return points in either counter/clockwise direction.

Reply:
I see! So, basically we would test for one or the other (A or B)? Are there any other possible configurations the points could be in?

Reply:
Not that I think of.. you can describe a polygon vertices either clockwise or counterclockwise from a given point, but any other order would be wrong

i.e. if you assume a pentagon has the vertices going A > B > C > D > E, if you were to order them as ACEBD they would be the same 5 points, but it would be a star, and not a convex polygon

So having ABCDEABCDE you can test for all 5 starting point possible clockwise sequences, and ECDBAECDBA for the counterclockwise ones

Reply:
This would be fairly trivial to test in Python, I think. We could create one array and rotate it for all possible rotations.